docs/FoLiA-hocr.1
.TH FoLiA-hocr 1 "2020 jan 19"

.SH NAME
FoLiA-hocr - Convert HOCR files into FoLiA.

.SH SYNOPSIS
FoLiA-hocr [options] FILE

FoLiA-hocr [options] DIR

.SH DESCRIPTION

When a DIR is provided,
.B FoLiA-hocr
will process all files in DIR and store the result in DIR, or in
the directory provided with
.B -O

When a FILE is provided,
.B FoLiA-hocr
will process that file and store its result in the directory where FILE is
found or in the directory provided with
.B -O

All files should be either HTML (.html), XHTML (.xhtml) or BZIP2 or GZ versions
of (X)HTML.
When the input file(s) are zipped, the output will be too.

.SH OPTIONS
.B --compress
kind
.RS
Use BZIP2 compression (kind=b) or GZIP compression (kind=g)
.RE

.B --setname
set
.RS
The FoLiA setname of the <str> nodes that are created. (default FoLiA-hocr-set).
.RE

.B --class
classname
.RS
The FoLiA classname of the <t> nodes that are created. (default OCR).
.RE

.B -O
prefix
.RS
use 'prefix' as a directory name to store the hocr-ed files in.
.RE

.B -t
or
.B --threads
number
.RS
Number of concurrent threads to be used by the program.
.RE

.B -v
.RS
be more verbose.
.RE

.B -V
or
.B --version
.RS
Show VERSION
.RE

.B -h
or
.B --help
.RS
Show some help
.RE

.SH BUGS
possible

.SH AUTHORS
Ko van der Sloot

Martin Reynaert

email: lamasoftware@science.ru.nl


NEWS
foliautils 0.16  07-01-2021
[Ko vd Sloot]
* requires libfolia 2.7 or above
* provenance data is better for a lot of modules
* added better checking on invalid NCnames in some modules.
* FoLiA-abby:
  - a lot of refactoring and additions to handle font/style information
* FoLiA-pm:
  - Notes are handled correctly now
  - fixed error in xlink attributes
* FoLiA-page:
  - more types of Page files are handled now
  - fixed annotation declarations
  - fixed offset calculation (due to change in FoLiA's opinion on those)
  - page number is added as a <br> node and in the metadata
  - added a --trusttokens option. This means that Word items in the Page file
    are added as Word's in the FoLiA, embedded in Sentences.
  - added a --norefs option to avoid adding references to the original texts
* FoLiA-correct:
  - make sure that the default is to run on 1 thread
  - added a --rebase-inputclass option

* FoLiA-alto:
  - the -t option was not always handled correctly

[Maarten van Gompel]
* FoLiA-benchmark: guard against compiler optimisation #48

foliautils 0.15  15-09-2020
[Maarten van Gompel]
* FoLiA-txt: check if a string is empty after normalisation (fix for #46)
[Ko vd Sloot]
* folia-correct: fix one-off error in hemp handling (when no hemp was found) #45
* some refactoring
  - centralized definition of XML_PARSER_OPTIONS
* bugfix in threading

foliautils 0.14  15-04-2020
[Martin Reynaert]
* updated man pages

[Ko vd Sloot]
* added man pages
* revised usage() in many modules
* the default seperator in FoLiA-stats is '_' now
* fix for: https://github.com/LanguageMachines/foliautils/issues/37
* fix for: https://github.com/LanguageMachines/foliautils/issues/41
* adapted to changes in libfolia
* many small code refactorings
* FoLiA-correct is improved a lot, allowing ngram corrections in FoLiA
* FoLiA-stats accepts a 'word_in_doc' mode now
* FoLiA-alto by default ceated <w> nodes now. use --oldstring to get <str>
* improved a lot in tests/
* many small fixes

foliautils 0.13  30-08-2019
[Ko vd Sloot]
bug fixes:
* fix for https://github.com/LanguageMachines/foliautils/issues/35
* fix for https://github.com/LanguageMachines/foliautils/issues/36

[Maarten van Gompel]
new features:
* FoLiA-wordtranslate.cxx: frog should be able to deal with
  spaces now, no need for ugly non-breaking space hack that now causes
  other problems in frog (LanguageMachines/frog#77)
* FoLiA-wordtranslate.cxx: adding ability to constrain by language


foliautils 0.12  29-05-2019
[Ko vd Sloot]
Release for FoLiA 2.0.

foliautils 0.11  15-05-2019
[Ko vd Sloot]
Stabilizing release for pre FoLiA 2.0

* Updated and added some tests
* started moving common code to a separate file and build a library
  (libfoliautils)
  - hemp detection is one of them

* FoLiA-stats:
  - added possibility to read a list of directories + file-names to process
    into seperate output directories. (could be generalized to other programs)
  - better hemp detection
* FoLiA-correct:
  - use same hempo detections as FoLiA-stats
* FoLiA-abby:
  - support more flavors
* FoLiA-clean:
  - avoid removing the last remaining tekt on <w> nodes
  - cleaning of tokenization now works


foliautils 0.10  29-11-2018
[Ko vd Sloot]
  * fixed icu:namespace issues
  * added FoLiA-abby, an ABBY to FoLiA convertor
  * src/FoLiA-abby.cxx, src/FoLiA-page.cxx, src/FoLiA-pm.cxx:
     - Allow 'none' value for --prefix
  * src/FoLiA-page.cxx, src/FoLiA-hocr.cxx: fixed Alignment info
  * src/FoLiA-correct.cxx:
     - fixed a problem with correction of the last word of a trigram.
     - fix correction of paragraphs with only deeper text
     - The --rank option accepts more flavors of files
  * src/FoLiA-stats.cxx:
     - added a --detokenize option
  * several minor fixes, refactorings etc.
  * updated tests

foliautils 0.9.2  05-06-2018
[Ko vd Sloot]
Bug fix release:
 * append small prefixes to output filenames, to ALWAYS avoid names starting with
   a numeric value.
   'FPM-' for FoLiA-pm. 'FP-' for FoLiA-page, 'FH-' for FoLiA-hocr
   Can bet set witth --prefix
 * FoLiA-stats.cxx:
   - added --collect to usage() and 'man' page
 * FoLiA-correct:
   - added --inputclass and --outputclass parameters (must be different)
   - Don't crash on empty text.

foliautils 0.9.1  17-05-2018
[Ko vd Sloot]
Bug fix release: the tests directory wasn't included in the distribution

foliautils 0.9  16-05-2018
[Ko vd Sloot]
* FoLiA-stats.cxx:
  - added a --collect option, to create files with all n-grams
    together
  - clearer message in FoLiA-stats when no results were found
  - extract text from deeper nodes, if needed
  - fixed out-of-bounds problem
  - now fails when every inputfile fails
* FoLiA-txt:
  - now fails when every inputfile fails
* avoid xml:id's starting with a number. Add "id-" in front.
* added more tests

[Maarten van Gompel]
* added codemeta.json


foliautils 0.8  19-02-2018
[Ko vd Sloot]
* added -R option to FoLiA-collect
* FoLiA-collect now can work in parallel (-t option)
* modernized configuration, whit better Max OSX support (including OpenMP)
* all modules end with an exit code now.
* added more tests to 'make check'
* added output of Type-Token Ratio's (also in degrees)
* several bugfixes.
* code cleanup and refactoring, some speedup too

foliautils 0.7  24-10-2017
[ko vd Sloot]
 * updated and expanded tests
 * fixed offset calculations in FoLiA-hocr, FoLiA-page.cxx
   and FoLiA-alto. We use unicode points now. (needed for folia v1.5 and above)
 * Changed 'modes' in FoLiA-stats, to be a bit more comprehensible
 * fixed problem with metadatatype when 'foreign-data' is present
 * enhanced FoLiA-clean. Still not done...
 * switched to dynamic OMP scheduling in most programs.
   (which process files with probably big differences in processing time)
 * small bugfixes.
 * general cleanup and refactoring

[Maarten van Gompel]
 * Added and improved FoLiA-wordtranslate.cxx

foliautils 0.6  04-04-2017
This is an intermediate release!!

src/FoLiA-hocr.cxx
/*
  Copyright (c) 2014 - 2021
  CLST  - Radboud University
  ILK   - Tilburg University

  This file is part of foliautils

  foliautils is free software; you can redistribute it and/or modify
  it under the terms of the GNU General Public License as published by
  the Free Software Foundation; either version 3 of the License, or
  (at your option) any later version.

  foliautils is distributed in the hope that it will be useful,
  but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
  GNU General Public License for more details.

  You should have received a copy of the GNU General Public License
  along with this program; if not, see <http://www.gnu.org/licenses/>.

  For questions and suggestions, see:
      https://github.com/LanguageMachines/foliautils/issues
  or send mail to:
      lamasoftware (at ) science.ru.nl
*/

#include <string>
#include <list>
#include <vector>
#include <iostream>
#include <fstream>
#include "ticcutils/StringOps.h"
#include "libfolia/folia.h"
#include "ticcutils/XMLtools.h"
#include "ticcutils/StringOps.h"
#include "ticcutils/zipper.h"
#include "ticcutils/FileUtils.h"
#include "ticcutils/Unicode.h"
#include "ticcutils/CommandLine.h"
#include "foliautils/common_code.h"
#include "config.h"
#ifdef HAVE_OPENMP
#include "omp.h"
#endif

using namespace	std;
using namespace	icu;

bool verbose = false;
string setname = "FoLiA-hocr-set";
string classname = "OCR";
string processor_id;

string extractContent( xmlNode* pnt ) {
  string result;
  if ( pnt ){
    result = TiCC::XmlContent(pnt);
    if ( result == "" )
      return extractContent( pnt->children );
  }
  return result;
}


void processParagraphs( xmlNode *div, folia::FoliaElement *out, const string& file ){
  list<xmlNode*> pars = TiCC::FindNodes( div, "//p" );
  if ( verbose ){
#pragma omp critical
    {
      cout << "processParagraphs: number=" << pars.size() << endl;
    }
  }
  for ( const auto& p : pars ){
    list<xmlNode*> lines = TiCC::FindNodes( p, ".//span[@class='ocr_line']" );
    if ( lines.empty() ){
#pragma omp critical
      {
	cerr << "found no 'ocr_line' nodes in paragraph id="
	     << TiCC::getAttribute( p, "id" ) << " (continuing)" << endl;
      }
      continue;
    }
    if ( verbose ){
#pragma omp critical
      {
	cout << "\thandling 'ocr_line' total=" << lines.size() << endl;
      }
    }
    string p_id = TiCC::getAttribute( p, "id" );
    folia::KWargs args;
    args["processor"] = processor_id;
    out->doc()->declare( folia::AnnotationType::PARAGRAPH, setname,  args );
    args.clear();
    args["xml:id"] = out->id() + "." + p_id;
    folia::Paragraph *par = new folia::Paragraph( args, out->doc() );
    UnicodeString txt;
    for ( const auto& line : lines ){
      list<xmlNode*> words = TiCC::FindNodes( line,
					      ".//span[@class='ocrx_word']" );
      if ( words.empty() ){
	// no ocrx_words. Lets see...
	words = TiCC::FindNodes( line, ".//span[@class='ocr_word']" );
	if ( words.empty() ){
#pragma omp critical
	  {
	    cerr << "found no OCRX_WORD or OCR_WORD nodes in " << file << endl;
	  }
	  delete par;
	  // maybe just warn and skip to next line?
	  return;
	}
      }
      if ( verbose ){
#pragma omp critical
	{
	  cout << "\t\thandling 'ocr_word' total=" << words.size() << endl;
	}
      }
      for ( const auto& word : words ){
	string w_id = TiCC::getAttribute( word, "id" );
	string content = extractContent( word );
	content = TiCC::trim( content );
	if ( !content.empty() ){
	  folia::KWargs args;
	  args["xml:id"] = par->id() + "." + w_id;
	  folia::String *str = new folia::String( args, out->doc() );
	  par->append( str );
	  UnicodeString uc = TiCC::UnicodeFromUTF8( content );
	  str->setutext( uc, txt.length(), classname );
	  txt += " " + uc;
	  args.clear();
	  args["processor"] = processor_id;
	  out->doc()->declare( folia::AnnotationType::RELATION,
			       setname,  args );
	  args.clear();
	  args["xlink:href"] = file;
	  args["format"] = "text/hocr+xml";
	  folia::Relation *h = new folia::Relation( args );
	  str->append( h );
	  args.clear();
	  args["id"] = w_id;
	  args["type"] = "str";
	  folia::LinkReference *a = new folia::LinkReference( args );
	  h->append( a );
	}
      }
    }
    if ( txt.length() > 1 ){
      out->append( par );
      par->setutext( txt.tempSubString(1), classname );
    }
    else {
      delete par;
    }
  }
}

string getDocId( const string& title ){
  string result;
  vector<string> vec = TiCC::split_at( title, ";" );
  for ( const auto& part : vec ){
    vector<string> v2 = TiCC::split( part );
    if ( v2.size() == 2 ){
      if ( TiCC::trim( v2[0] ) == "image" ){
	result = v2[1];
	string::size_type pos = result.rfind( "/" );
	if ( pos != string::npos ){
	  result = result.substr( pos+1 );
	}
      }
    }
  }
  result = TiCC::trim( result, " \t\"" );
  return result;
}

void convert_hocr( const string& fileName,
		   const string& outputDir,
		   const zipType outputType,
		   const string& prefix,
		   const string& command ){
  if ( verbose ){
#pragma omp critical
    {
      cout << "start handling " << fileName << endl;
    }
  }
  zipType inputType;
  xmlDoc *xdoc = getXml( fileName, inputType );
  if ( !xdoc ){
#pragma omp critical
    {
      cerr << "problem detecting type of file: " << fileName << endl;
      cerr << "it MUST have extension .hocr, .html or .xhtml (or .bz2 or .gz variants)" << endl;
    }
    return;
  }
  xmlNode *root = xmlDocGetRootElement( xdoc );
  list<xmlNode*> divs = TiCC::FindNodes( root, "//div[@class='ocr_page']" );
  if ( divs.empty() ) {
#pragma omp critical
    {
      cerr << "no OCR_PAGE node found in " << fileName << endl;
    }
    exit(EXIT_FAILURE);
  }
  if ( divs.size() > 1 ) {
#pragma omp critical
    {
      cerr << "multiple OCR_PAGE nodes found. Not supported. in " << fileName << endl;
    }
    exit(EXIT_FAILURE);
  }
  string title = TiCC::getAttribute( *divs.begin(), "title" );
  if ( title.empty() ){
#pragma omp critical
    {
      cerr << "No 'title' attribute found in ocr_page: " << fileName << endl;
    }
    exit(EXIT_FAILURE);
  }
  title = getDocId( title );
  string docid = title.substr( 0, title.find(".") );
  docid = prefix + docid;
  folia::Document doc( "xml:id='" + docid + "'" );
  doc.set_metadata( "filename", title );
  folia::processor *proc = add_provenance( doc, "FoLiA-hocr", command );
  processor_id = proc->id();
  folia::KWargs args;
  args["processor"] = processor_id;
  doc.declare( folia::AnnotationType::STRING, setname,  args );
  args.clear();
  args["xml:id"] = docid + ".text";
  folia::Text *text = new folia::Text( args );
  doc.append( text );
  processParagraphs( root, text, TiCC::basename(fileName) );
  xmlFreeDoc( xdoc );

  string outName = outputDir + "/" + prefix + title + ".folia.xml";
  zipType type = inputType;
  if ( outputType != NORMAL ){
    type = outputType;
  }
  if ( type == BZ2 ){
    outName += ".bz2";
  }
  else if ( type == GZ ){
    outName += ".gz";
  }
  vector<folia::Paragraph*> pv = doc.paragraphs();
  if ( pv.size() == 0 ||
       ( pv.size() == 1 && pv[0]->size() == 0 ) ){
    // no paragraphs, or just 1 without data
#pragma omp critical
    {
      cerr << "skipped empty result : " << outName << endl;
    }
  }
  else {
    doc.save( outName );
    if ( verbose ){
#pragma omp critical
      {
	cout << "created " << outName << endl;
      }
    }
  }
}

void usage(){
  cerr << "Usage: FoLiA-hocr [options] file/dir" << endl;
  cerr << "\t-t 'threads' or\n\t--threads='threads' Number of threads to run on." << endl;
  cerr << "\t\t\t If 'threads' has the value \"max\", the number of threads is set to a" << endl;
  cerr << "\t\t\t reasonable value. (OMP_NUM_TREADS - 2)" << endl;
  cerr << "\t-h or --help\t this message " << endl;
  cerr << "\t-O\t\t output directory " << endl;
  cerr << "\t--compress='c'\t with 'c'=b create bzip2 files (.bz2) " << endl;
  cerr << "\t\t\t with 'c'=g create gzip files (.gz)" << endl;
  cerr << "\t--setname='set'\t the FoLiA set name for <t> nodes. "
    "(default '" << setname << "')" << endl;
  cerr << "\t--class='class'\t the FoLiA class name for <t> nodes. "
    "(default '" << classname << "')" << endl;
  cerr << "\t--prefix='pre'\t add this prefix to ALL created files. (default 'FH-') " << endl;
  cerr << "\t\t\t use 'none' for an empty prefix. (can be dangerous)" << endl;
  cerr << "\t-v\t\t verbose output " << endl;
  cerr << "\t-V or --version\t show version " << endl;
}

int main( int argc, char *argv[] ){
  TiCC::CL_Options opts( "vVt:O:h",
			 "compress:,class:,setname:,help,version,prefix:,threads:" );
  try {
    opts.init( argc, argv );
  }
  catch( TiCC::OptionError& e ){
    cerr << e.what() << endl;
    usage();
    exit( EXIT_FAILURE );
  }
#ifdef HAVE_OPENMP
  int numThreads=1;
#endif

  string outputDir;
  zipType outputType = NORMAL;
  string value;
  if ( opts.empty() ){
    usage();
    exit(EXIT_FAILURE);
  }
  if ( opts.extract( 'h' ) || opts.extract( "help" ) ){
    usage();
    exit(EXIT_SUCCESS);
  }
  if ( opts.extract( 'V' ) || opts.extract( "version" ) ){
    cerr << PACKAGE_STRING << endl;
    exit(EXIT_SUCCESS);
  }
  string orig_command = "FoLiA-hocr " + opts.toString();
  verbose = opts.extract( 'v', value );
  if ( opts.extract( "compress", value ) ){
    if ( value == "b" )
      outputType = BZ2;
    else if ( value == "g" )
      outputType = GZ;
    else {
      cerr << "unknown compression: use 'b' or 'g'" << endl;
      exit( EXIT_FAILURE );
    }
  }
  if ( opts.extract( 't', value )
       || opts.extract( "threads", value ) ){
#ifdef HAVE_OPENMP
    if ( TiCC::lowercase(value) == "max" ){
      numThreads = omp_get_max_threads() - 2;
    }
    else if ( !TiCC::stringTo(value,numThreads) ) {
      cerr << "illegal value for -t (" << value << ")" << endl;
      exit( EXIT_FAILURE );
    }
#else
    cerr << "You don't have OpenMP support, so '-t' option is impossible" << endl;
    exit(EXIT_FAILURE);
#endif
  }
  opts.extract( 'O', outputDir );
  opts.extract( "setname", setname );
  opts.extract( "class", classname );
  string prefix = "FH-";
  opts.extract( "prefix", prefix );
  if ( prefix == "none" ){
    prefix.clear();
  }
  vector<string> fileNames = opts.getMassOpts();
  if ( fileNames.empty() ){
    cerr << "missing input file(s)" << endl;
    exit(EXIT_FAILURE);
  }
  else if ( fileNames.size() > 1 ){
    cerr << "currently only 1 file or directory is supported" << endl;
    exit( EXIT_FAILURE );
  }

  if ( !outputDir.empty() ){
    string name = outputDir;
    if ( !TiCC::isDir(name) ){
      if ( !TiCC::createPath( name ) ){
	cerr << "outputdir '" << name
	     << "' doesn't exist and can't be created" << endl;
	exit(EXIT_FAILURE);
      }
    }
  }
  string name = fileNames[0];
  if ( !( TiCC::isFile(name) || TiCC::isDir(name) ) ){
    cerr << "parameter '" << name << "' doesn't seem to be a file or directory"
	 << endl;
    exit(EXIT_FAILURE);
  }
  if ( TiCC::isFile(name) ){
    if ( TiCC::match_back( name, ".tar" ) ){
      cerr << "TAR files are not supported yet." << endl;
      exit(EXIT_FAILURE);
    }
  }
  else {
    fileNames = TiCC::searchFilesMatch( name, "*hocr" );
    if ( fileNames.empty() ){
      fileNames = TiCC::searchFilesMatch( name, "*html" );
    }
  }
  size_t toDo = fileNames.size();
  if ( toDo == 0 ){
    cerr << "no matching files found." << endl;
    exit(EXIT_FAILURE);
  }
  if ( toDo > 1 ){
    cout << "start processing of " << toDo << " files " << endl;
  }

#ifdef HAVE_OPENMP
  if ( numThreads >= 1 ){
    omp_set_num_threads( numThreads );
  }
#endif

#pragma omp parallel for shared(fileNames) schedule(dynamic)
  for ( size_t fn=0; fn < fileNames.size(); ++fn ){
    convert_hocr( fileNames[fn], outputDir, outputType, prefix, orig_command );
  }
  cout << "done" << endl;
  exit(EXIT_SUCCESS);
}
